File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/MaxSumOfThreeConsecutive.c
#include "java_arrays.h"
#include <stdint.h>

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left - right);
*/
static int32_t java_sub(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left - (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (left > right ? left : right);
*/
static int32_t java_max(int32_t left, int32_t right) {
    return left > right ? left : right;
}

/*@
  logic integer fb_three{L}(JIntArray a, integer k) =
    k <= 0 ? a->data[0] : k == 1 ? (int32_t)((integer)a->data[0] + a->data[1]) :
    k == 2 ? \max(fb_three(a, 1), \max((int32_t)((integer)a->data[1] + a->data[2]), (int32_t)((integer)a->data[0] + a->data[2]))) :
    \max(\max(fb_three(a, k - 1), (int32_t)(fb_three(a, k - 2) + a->data[k])),
         (int32_t)((integer)a->data[k] + a->data[k - 1] + fb_three(a, k - 3)));
*/
/*@
  requires jintarray_valid_nonnull(arr) && 1 <= n <= arr->length;
  assigns \nothing;
  ensures \result == fb_three(arr, n - 1);
*/
int32_t maxSumOfThreeConsecutive(JIntArray arr, int32_t n) {
    JIntArray sum = jarray_new(n);
    if (n >= 1) {
        jarray_set(sum, 0, jarray_get(arr, 0));
    }
    if (n >= 2) {
        jarray_set(sum, 1, java_add(jarray_get(sum, 0), jarray_get(arr, 1)));
    }
    if (n > 2) {
        int32_t candidate1 = jarray_get(sum, 1);
        int32_t candidate2 = java_add(jarray_get(arr, 1), jarray_get(arr, 2));
        int32_t candidate3 = java_add(jarray_get(arr, 0), jarray_get(arr, 2));
        jarray_set(sum, 2, java_max(candidate1, java_max(candidate2, candidate3)));
    }
    /*@
      loop invariant 3 <= i && (n < 3 ? i == 3 : i <= n);
      loop invariant jintarray_valid_nonnull(sum) && sum->length == n;
      loop invariant \forall integer k; 0 <= k < i && k < n ==> sum->data[k] == fb_three(arr, k);
      loop assigns i, sum->data[0 .. n - 1];
      loop variant (integer)n - i;
    */
    for (int32_t i = 3; i < n; i = java_add(i, 1)) {
        int32_t first = jarray_get(sum, java_sub(i, 1));
        int32_t second = java_add(jarray_get(sum, java_sub(i, 2)), jarray_get(arr, i));
        int32_t third = java_add(java_add(jarray_get(arr, i), jarray_get(arr, java_sub(i, 1))),
                                 jarray_get(sum, java_sub(i, 3)));
        jarray_set(sum, i, java_max(java_max(first, second), third));
    }
    return jarray_get(sum, java_sub(n, 1));
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/MaxSumOfThreeConsecutive.java

import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;

class MaxSumOfThreeConsecutive {
    
    /*@ public normal_behavior
      @ ensures \result == ((x % 4294967296L + 6442450944L) % 4294967296L) - 2147483648L;
      @ model public static pure \bigint fb_wrap(\bigint x);
      @*/
    /*@ public normal_behavior
      @ ensures \result == (a >= b ? a : b);
      @ model public static pure \bigint fb_max(\bigint a, \bigint b);
      @*/
    /*@ public normal_behavior
      @ requires a != null && 0 <= k < a.length;
      @ ensures \result == (k == 0 ? a[0] : k == 1 ? fb_wrap((\bigint)a[0] + a[1]) : k == 2 ? fb_max(fb_three(a, 1), fb_max(fb_wrap((\bigint)a[1] + a[2]), fb_wrap((\bigint)a[0] + a[2]))) : fb_max(fb_max(fb_three(a, k - 1), fb_wrap(fb_three(a, k - 2) + a[k])), fb_wrap((\bigint)a[k] + a[k - 1] + fb_three(a, k - 3))));
      @ measured_by k;
      @ model public static pure \bigint fb_three(int[] a, int k);
      @*/
    /*@ public normal_behavior
      @ requires arr != null && 1 <= n <= arr.length;
      @ assignable \nothing;
      @ ensures \result == fb_three(arr, n - 1);
      @ also
      @ public exceptional_behavior
      @ requires n < 0;
      @ assignable \nothing;
      @ signals_only NegativeArraySizeException;
      @ also
      @ public exceptional_behavior
      @ requires n == 0 || (n > 0 && arr != null && n > arr.length);
      @ assignable \nothing;
      @ signals_only ArrayIndexOutOfBoundsException;
      @ also
      @ public exceptional_behavior
      @ requires n > 0 && arr == null;
      @ assignable \nothing;
      @ signals_only NullPointerException;
      @*/
    //@ code_java_math
    public static int maxSumOfThreeConsecutive(int[] arr, int n) {
        int[] sum = new int[n];
        if (n >= 1) {
            sum[0] = arr[0];
        }
        if (n >= 2) {
            sum[1] = sum[0] + arr[1];
        }
        if (n > 2) {
            sum[2] = Math.max(sum[1], Math.max(arr[1] + arr[2], arr[0] + arr[2]));
        }
        /*@ loop_invariant 3 <= i && (n < 3 ? i == 3 : i <= n);
          @ loop_invariant (\forall int k; 0 <= k < i && k < n; sum[k] == fb_three(arr, k));
          @ loop_writes i, sum[*];
          @ decreases (\bigint)n - i;
          @*/
        for (int i = 3; i < n; i++) {
            sum[i] = Math.max(Math.max(sum[i - 1], sum[i - 2] + arr[i]), arr[i] + arr[i - 1] + sum[i - 3]);
        }
        return sum[n - 1];
    }
}

